[16.0][MIG] pos_tare: migration to 16.0

huguesdk opened the pull request:
this is a follow-up on #1224. it contains all of its commits (as it’s the same branch), as well as several commits to restore features present in version 12.0 but that disappeared during the migration to 14.0, with also some fixes and improvements.

legalsylvain committed:
[ADD][8.0] pos_tare

legalsylvain committed:
[MIG] pos_tare: Migration to 12.0
[ADD] barcode nomenclature for tare
[ADD] tare field on pos.order.line model
[ADD] Tare button in PoS numpad
[IMP] handle correctly different UoM
[IMP] add warning at checkout if qty <=0
[IMP] display gross and tare weight for each line

Co-authored-by: François Kawala <francois.kawala@gmail.com>
Co-authored-by: Sylvain LE GAL <sylvain.legal@grap.coop>

legalsylvain committed:
[REF] pos_tare: Black python code

oca-travis committed:
[UPD] Update pos_tare.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
pos_tare 12.0.1.0.1

danimv5 committed:
Added translation using Weblate (Spanish)

danimv5 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (46 of 46 strings)

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/es/

oca-travis committed:
[UPD] Update pos_tare.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/

legalsylvain committed:
[FIX] pos_tare : raise an error if the gross weight is not correct

legalsylvain committed:
[FIX] pos_tare : move the confirm popup regarding null or negative quantity in the correct place

oca-travis committed:
[UPD] Update pos_tare.pot

OCA-git-bot committed:
pos_tare 12.0.1.0.2

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/

oca-travis committed:
[UPD] Update pos_tare.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/

legalsylvain committed:
[IMP][12.O] pos_tare : add default tare weight on product.template model

oca-travis committed:
[UPD] Update pos_tare.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
pos_tare 12.0.1.0.3

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/

oca-travis committed:
[UPD] Update pos_tare.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (55 of 55 strings)

Translation: pos-12.0/pos-12.0-pos_tare
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/it/

baimont committed:
[IMP] pos_tare: black, isort, prettier

baimont committed:
[MIG] pos_tare: Migration to 14.0

robinkeunen committed:
[REF] pos_tare: do not rewrite _getAddProductOptions

robinkeunen committed:
[IMP] pos_tare: black, isort, prettier

robinkeunen committed:
[MIG] pos_tare: Migration to 16.0

huguesdk committed:
[IMP] improve readme

huguesdk committed:
[FIX] re-add lost barcode error handling code

huguesdk committed:
[FIX] fix english in a field help string

huguesdk committed:
[FIX] fix loading of tare at order line creation

huguesdk committed:
[FIX] fix numpad layout css

huguesdk committed:
[IMP] convert to native javascript modules

huguesdk committed:
[FIX] add missing units to receipt lines

huguesdk committed:
[FIX] restore numpad button positions

huguesdk committed:
[FIX] re-add lost weight fields value validation

huguesdk committed:
[FIX] fix field help string

huguesdk committed:
[IMP] remove useless installable property

huguesdk committed:
[IMP] improve scale screen input

* set placeholders to 0 instead of 0.000, as a 0 value is formatted like
  that.
* set fields empty when the value is 0, to make manual input easier (no
  need to select or delete the existing value first).
* focus on the first editable field and select its input.
* evaluate the input values at each input event (as was done previously)
  instead of when leaving the field.

huguesdk committed:
[IMP] re-add lost red background on invalid entry

re-add lost red background on invalid entry in the scale screen fields.

huguesdk committed:
[FIX] restore lost behavior of quantity input

make sure inputting a quantity sets the gross weight (like how it was
previously) instead of the net weight.

huguesdk committed:
[FIX] use default product tare without scale

when the scale is disabled, ensure that the default product tare is also
set on the order line.

huguesdk committed:
[FIX] only set up barcode action when needed

enable or disable barcode action according to the value of
iface_tare_method (like it was done previously).

huguesdk committed:
[IMP] handle barcode scanning on the scale screen

huguesdk committed:
[IMP] improve scale screen

* always display the gross and tare weight values regardless of gross
  and tare weight input methods, but make them editable or not depending
  on these input methods.
* improve style: make it more consistent by reusing existing css classes
  and adapt them where needed.
* automatically focus on the first editable field.
* display the correct gross weight uom.

huguesdk committed:
[FIX] correctly use uom on scale screen

compute the net weight correctly on the scale screen by taking the
product and tare weight uom into account and converting between them if
needed.

huguesdk committed:
[IMP] update contributors

huguesdk committed:
[IMP] simplify convert_mass()

huguesdk committed:
[FIX] fix fr translation

huguesdk committed:
[FIX] don't focus fields when using barcode reader

don't focus fields on the scale screen when the tare can be input with a
barcode reader to avoid filling the field with the barcode value.

victor-champonnois committed:
[IMP] update and improve documentation

victor-champonnois commented:
@huguesdk I've updated the doc but the image links do not work, I don't manage to figure out why.

huguesdk commented:
@victor-champonnois thanks! it's normal: images link to their location on github in the main branch for the version (`16.0` in this case). they will appear only when the branch will be merged.

huguesdk commented:
@OCA/pos-maintainers could anyone please review or merge this?

polchampion commented:
@OCA/pos-maintainers could anyone please review or merge this?

Fkawala commented:
LGTM, but I'd like someone more active in the project @OCA/pos-maintainers to have a second look on this before merging. Thank you so much for your work on this @huguesdk !

ivantodorovich reviewed (approved):
I don't have experience with this module, but technically it looks good.
I would perhaps encourage you to squash some of the commits into the migration one, as they seem to be a part of it

ferdymercury commented on pos_tare/README.rst:
```diff
@@ -0,0 +1,122 @@
+.. figure:: https://raw.githubusercontent.com/OCA/pos/16.0/pos_tare/static/description/pos_config.png
+
+
+To generate a tare barcode you need to use the default barcode nomenclature. The default barcode pattern is `0700000{NNDDD}`. Using that pattern, the barcode for a tare of 0.1kg is `0700000001006`. The `pos_tare_barcode_generator` allows you to create tare labels right from the POS.
```
```suggestion
To generate a tare barcode you need to use the [default barcode nomenclature](https://www.odoo.com/documentation/16.0/applications/inventory_and_mrp/barcode/operations/barcode_nomenclature.html). The default EAN-13 barcode pattern is `0700000{NNDDD}C`, where `NN` are the kilograms, `DDD` are the grams and `C` is a check digit. Using that pattern, the barcode for a tare of 0.123kg is `0700000001236`. The maximum weight that can be represented with this barcode is just below 100 kg. The embedded barcode generator allows you to create and print tare labels right from the POS.
```

Is pos_tare_barcode_generator a separate module or subfunction ? If not, maybe reword as suggested.

huguesdk replied:
i’m not aware of this module. i quickly looked in earlier versions but did not find it. another module ([``pos_self_service_weighing_tare``](https://github.com/coopiteasy/pos/tree/16.0-self-service-weighing/pos_self_service_weighing_tare), see #1397) allows to do this (but not from another pos than the one used to sell products).

ferdymercury commented on pos_tare/README.rst:
```diff
@@ -0,0 +1,122 @@
+
+.. figure:: https://raw.githubusercontent.com/OCA/pos/16.0/pos_tare/static/description/pos_config.png
+
+
```
One curiosity, why did you choose 07 as prefix?

huguesdk replied:
i don’t know where this prefix comes from. @legalsylvain do you know this?

ferdymercury commented on pos_tare/README.rst:
```diff
@@ -0,0 +1,122 @@
+
+To generate a tare barcode you need to use the default barcode nomenclature. The default barcode pattern is `0700000{NNDDD}`. Using that pattern, the barcode for a tare of 0.1kg is `0700000001006`. The `pos_tare_barcode_generator` allows you to create tare labels right from the POS.
+
+you can define a default tare on the product form view, if you always use the same type of packaging (or container) for a given product.
```
```suggestion
You can define a default tare on the product form view, if you always use the same type of packaging (or container) for a given product.
```

ferdymercury commented on pos_tare/readme/USAGE.rst:
```diff
@@ -0,0 +1,14 @@
+.. figure:: ../static/description/pos_config.png
+
+
+To generate a tare barcode you need to use the default barcode nomenclature. The default barcode pattern is `0700000{NNDDD}`. Using that pattern, the barcode for a tare of 0.1kg is `0700000001006`. The `pos_tare_barcode_generator` allows you to create tare labels right from the POS.
```
Same comments/corrections as above would apply, since this paragraph is duplicated.

Maybe this whole file can be replaced with just a link pointing to the proper section in README.rst ?

ferdymercury commented on pos_tare/README.rst:
```diff
@@ -0,0 +1,122 @@
+Give the possibility to the user to enter a Tare weight, when weighing
+products in the Point of Sale.
+This will compute automatically net weight and set it to the current
+selected order
```
```suggestion
selected order.
```

ferdymercury commented on pos_tare/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,3 @@
+- Sylvain LE GAL (https://www.twitter.com/legalsylvain)
```
in the README.rst, also GRAP is listed

huguesdk replied:
sylvain le gal works for grap.

ferdymercury commented on pos_tare/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,15 @@
+Give the possibility to the user to enter a Tare weight, when weighing
+products in the Point of Sale.
+This will compute automatically net weight and set it to the current
+selected order
```
```suggestion
selected order.
```

Or remove with a link to README to avoid duplication?

ferdymercury reviewed (changes requested):
Thanks a lot for this migration.

Looks good to me for merging. Just indicating some irrelevant typos and clarifications that could be fixed.

ferdymercury commented on pos_tare/static/description/index.html:
```diff
@@ -0,0 +1,459 @@
+<img alt="https://raw.githubusercontent.com/OCA/pos/16.0/pos_tare/static/description/pos_config.png" src="https://raw.githubusercontent.com/OCA/pos/16.0/pos_tare/static/description/pos_config.png" />
+</div>
+<p>To generate a tare barcode you need to use the default barcode nomenclature. The default barcode pattern is <cite>0700000{NNDDD}</cite>. Using that pattern, the barcode for a tare of 0.1kg is <cite>0700000001006</cite>. The <cite>pos_tare_barcode_generator</cite> allows you to create tare labels right from the POS.</p>
+<p>you can define a default tare on the product form view, if you always use the same type of packaging (or container) for a given product.</p>
```
Same corrections / comments as above.

ferdymercury commented on pos_tare/static/description/index.html:
```diff
@@ -0,0 +1,459 @@
+<p>Give the possibility to the user to enter a Tare weight, when weighing
+products in the Point of Sale.
+This will compute automatically net weight and set it to the current
+selected order</p>
```
```suggestion
selected order.</p>
```

huguesdk commented:
@ivantodorovich thanks for the review. i’m not sure about squashing the commits, as many features of version 12.0 had been lost in the porting to 14.0, and they had to be re-added after the porting to 16.0. that’s why i think it is useful to keep them separate, especially as they explain better each change. what do you think?

huguesdk reviewed (commented):
@ferdymercury thanks for the review. i’ve update the readme with your suggestions (and improved it a little bit also). please note that the readme fragments (from the `readme` directory) are restructuredtext, not markdown, and that `README.rst` and `static/description/index.html` are automatically generated from them.

ferdymercury reviewed (approved):
Thanks a lot for the fixes!

I agree it's better to keep the current commit history (no need to squash)

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#. module: pos_tare
+#: selection:barcode.rule,type:0
+msgid "Meal Voucher Payment"
+msgstr ""
```
```suggestion
msgstr "Pago de cupón de comida"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: code:addons/pos_tare/static/src/js/screens.js:185
+#, python-format
+msgid "Feature Disabled"
+msgstr ""
```
```suggestion
msgstr "Característica deshabilitada"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: model:product.product,name:pos_tare.cheese
+#: model:product.template,name:pos_tare.cheese_product_template
+msgid "Cheese by the Slice"
+msgstr ""
```
```suggestion
msgstr "Queso al corte"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: code:addons/pos_tare/static/src/xml/pos_tare.xml:34
+#, python-format
+msgid "Gross Weight:"
+msgstr ""
```
```suggestion
msgstr "Peso bruto:"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: code:addons/pos_tare/static/src/js/screens.js:91
+#, python-format
+msgid "Incorrect Gross Weight Value"
+msgstr ""
```
```suggestion
msgstr "Valor de peso bruto incorrecto"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: code:addons/pos_tare/static/src/js/screens.js:92
+#, python-format
+msgid "Please set a numeric value in the gross weight field."
+msgstr ""
```
```suggestion
msgstr "Por favor introduzca un valor numérico en el campo de peso bruto"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#. module: pos_tare
+#: selection:barcode.rule,type:0
+msgid "Priced Product (Computed Weight)"
+msgstr "Producto con precio (precio calculado)"
```
```suggestion
msgstr "Producto con precio (peso calculado)"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#. module: pos_tare
+#: model:ir.model,name:pos_tare.model_product_template
+msgid "Product Template"
+msgstr ""
```
```suggestion
msgstr "Plantilla de producto"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+msgid ""
+"Select tare method:\n"
+"* 'manual' : the scale screen has an extra tare input field;\n"
+"* 'barecode' : (scan a barcode to tare the selected order line;\n"
```
```suggestion
"* 'barcode' : scan a barcode to tare the selected order line;\n"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+"Seleccionar método de tara:\n"
+"* 'manual': la pantalla de escala tiene un campo de entrada de tara "
+"adicional;\n"
+"* 'barecode': (escanea un código de barras para tarar la línea de orden "
```
```suggestion
"* 'código de barras': escanea un código de barras para tarar la línea de orden "
```

ferdymercury commented on pos_tare/i18n/fr.po:
```diff
@@ -0,0 +1,364 @@
+msgid ""
+"Select tare method:\n"
+"* 'manual' : the scale screen has an extra tare input field;\n"
+"* 'barecode' : (scan a barcode to tare the selected order line;\n"
```
```suggestion
"* 'barcode' : scan a barcode to tare the selected order line;\n"
```

ferdymercury commented on pos_tare/i18n/it.po:
```diff
@@ -0,0 +1,359 @@
+#: model:ir.model.fields,help:pos_tare.field_pos_config__iface_tare_method
+msgid "Select tare method:\n"
+"* 'manual' : the scale screen has an extra tare input field;\n"
+"* 'barecode' : (scan a barcode to tare the selected order line;\n"
```
```suggestion
"* 'barcode' : scan a barcode to tare the selected order line;\n"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+"adicional;\n"
+"* 'barecode': (escanea un código de barras para tarar la línea de orden "
+"seleccionada;\n"
+"* 'both': los métodos de entrada manual y de código de barras están "
```
```suggestion
"* 'ambos': los métodos de entrada manual y de código de barras están "
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+msgid ""
+"Set here Constant tare weight for the given product. This tare will be "
+"substracted when the product is weighted"
+msgstr ""
```
```suggestion
msgstr "Introduzca aquí una tara constante para el producto. Será "
"restada cuando se pese el producto"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: model:ir.model.fields,field_description:pos_tare.field_product_product__tare_weight
+#: model:ir.model.fields,field_description:pos_tare.field_product_template__tare_weight
+msgid "Tare Weight"
+msgstr ""
```
```suggestion
msgstr "Peso de tara"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#, python-format
+msgid ""
+"The quantity is lower or equal to zero. Are you sure you want to continue ?"
+msgstr ""
```
```suggestion
msgstr "Cantidad menor o igual a cero. ¿Seguro que desea continuar?"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#. module: pos_tare
+#: selection:barcode.rule,type:0
+msgid "Weighted Product"
+msgstr "Producto ponderado"
```
```suggestion
msgstr "Producto pesado"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#: model:product.template,uom_name:pos_tare.cheese_product_template
+#: model:product.template,weight_uom_name:pos_tare.cheese_product_template
+msgid "kg"
+msgstr ""
```
```suggestion
msgstr "kg"
```

ferdymercury commented on pos_tare/i18n/fr.po:
```diff
@@ -0,0 +1,364 @@
+"Selection la méthode de tare : \n"
+" * 'manuelle' : l'écran de pesée à un nouveau champ de saisie pour la tare;\n"
+" * 'barcode' : scanner un code bare pour tarer la ligne sélectionnée;\n"
+"* 'both : Saisie manuelle ou par code barre"
```
```suggestion
" * 'both : Saisie manuelle ou par code barre"
```

ferdymercury commented on pos_tare/i18n/fr.po:
```diff
@@ -0,0 +1,364 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 12.0\n"
```
```suggestion
"Project-Id-Version: Odoo Server 16.0\n"
```

ferdymercury commented on pos_tare/i18n/es.po:
```diff
@@ -0,0 +1,373 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 12.0\n"
```
```suggestion
"Project-Id-Version: Odoo Server 16.0\n"
```

ferdymercury commented on pos_tare/i18n/it.po:
```diff
@@ -0,0 +1,359 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 12.0\n"
```
```suggestion
"Project-Id-Version: Odoo Server 16.0\n"
```

ferdymercury commented on pos_tare/i18n/pos_tare.pot:
```diff
@@ -0,0 +1,346 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 12.0\n"
```
```suggestion
"Project-Id-Version: Odoo Server 16.0\n"
```

ferdymercury commented on pos_tare/i18n/pos_tare.pot:
```diff
@@ -0,0 +1,346 @@
+#: model:ir.model.fields,help:pos_tare.field_pos_config__iface_tare_method
+msgid "Select tare method:\n"
+"* 'manual' : the scale screen has an extra tare input field;\n"
+"* 'barecode' : (scan a barcode to tare the selected order line;\n"
```
```suggestion
"* 'barcode' : scan a barcode to tare the selected order line;\n"
```

ferdymercury reviewed (commented):
some minor typos in the translation

huguesdk commented:
@ferdymercury the translation files will be updated automatically by the bot after merge. once merged, you will be able to add new translations using [weblate](https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/) (link to version 12 of the translation of this module, as 16 will be created after merge).

huguesdk committed:
[FIX] warn when uom categories don't match

* display an error message when a tare is set on a product that uses a
  different uom category than the one used for the tare instead of just
  throwing an exception and making the pos unusable until it is
  reloaded.
* don't compute the weight if no tare is set to avoid uom categories
  mismatch errors, falling back to odoo's default behavior.

huguesdk commented:
i’ve added a new commit to gracefully handle uom categories mismatch errors.

ferdymercury commented:
> @ivantodorovich thanks for the review. i’m not sure about squashing the commits, as many features of version 12.0 had been lost in the porting to 14.0, and they had to be re-added after the porting to 16.0. that’s why i think it is useful to keep them separate, especially as they explain better each change. what do you think?

gentle ping @ivantodorovich
(thanks in advance)

ivantodorovich commented:
/ocabot migration pos_tare
/ocabot merge nobump

ferdymercury commented:
Thanks for the merge!

> you will be able to add new translations using [weblate](https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_tare/)

I have added suggestions now to weblate, thanks. Not sure who needs to approve them.

huguesdk commented:
@ferdymercury did you register on weblate as explained on [this page](https://odoo-community.org/resources/translate)? once you are accepted, you will be able to make changes directly, and they will be automatically applied (not directly, but periodically) to the source files here on github.

ferdymercury commented:
Nope, I submitted the suggestions as unlogged anonymous :S
(Thanks I will follow the login step now)
